pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -125,13 +125,64 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
 	// where files will land inside the new image.
 	scriptsDestDir := filepath.Join(getDestination(config), "scripts")
 	sourceDestDir := filepath.Join(getDestination(config), "src")
+	artifactsDestDir := filepath.Join(getDestination(config), "artifacts")
+	artifactsTar := sanitize(filepath.ToSlash(filepath.Join("/tmp", "artifacts.tar")))
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(config.WorkingDir, builder.uploadScriptsDir))
+	assembleProvided := false
+	runProvided := false
+	saveArtifactsProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found override script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		} else if f.Name() == "save-artifacts" {
+			saveArtifactsProvided = true
+		}
+		if runProvided && assembleProvided && saveArtifactsProvided {
+			break
+		}
+	}
+
+	if config.Incremental {
+		imageTag := util.FirstNonEmpty(config.IncrementalFromTag, config.Tag)
+		if len(imageTag) == 0 {
+			return errors.New("Image tag is missing for incremental build")
+		}
+		buffer.WriteString(fmt.Sprintf("FROM %s as cached\n", imageTag))
+		if len(config.AssembleUser) > 0 {
+			buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+		}
```

Comment:
I had the most concerns re: this instruction. `config.AssembleUser` is always non-empty because we default it to 1001. There is always a possibility that this user does not have permissions to run the `save-artifacts` script that is included in the image.

On the other hand, if the user provides an override for `save-artifacts` we need the `--chown` instruction with the `COPY` directive.

Reply:
> There is always a possibility that this user does not have permissions to run the save-artifacts script that is included in the image.

i think that possibility already exists in the normal s2i incremental flow today.